### PR TITLE
cdb2sql: autocomplete comdb2_systables and @ commands

### PR DESCRIPTION
**Pull request by adizaimi**
cdb2sql: autocomplete comdb2_systables and @ commands

### CODE REVIEW

File: tools/cdb2sql/cdb2sql.cpp [hunk 1]
```diff
@@ -161,7 +161,8 @@ void cdb2sql_usage(int exit_val)
 }
 
 
-const char *words[] = {
+const char *level_one_words[] = {
+  "@desc", "@ls", "@send",
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: Perhaps not show these until "@ + TAB" ?

File: tools/cdb2sql/cdb2sql.cpp [hunk 2]
```diff
@@ -175,8 +176,31 @@ const char *words[] = {
   "SELECT", "SELECTV", "SET",
   "TRUNCATE",
   "UPDATE",
-  "WITH", NULL, }; // must be terminated by NULL
+  "WITH", NULL,  // must be terminated by NULL
+};
+
+
+const char *char_alpha_words[] = {
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: Could you explain this array? The name's not very clear.

File: tools/cdb2sql/cdb2sql.cpp [hunk 3]
```diff
@@ -297,50 +321,72 @@ char *generic_generator(const char *text, int state)
...
 
 
 // Custom completion function
+//
+// text is the last word entered or space, 
+// start and end are the offsets of that last word in the rl_line_buffer
+// So if we are trying to complete in the middle of a word
+// text is the [partial] word, example: 
+//  > abra kadabr
+//   rl_line_buffer='abra kadabr' text='kadabr' start=5 end=11
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: I'd refrain from using such words, may be just use examples pertaining to comdb2/cdb2sql?

File: sqlite/ext/misc/completion.c [hunk 4]
```diff
@@ -242,10 +242,10 @@ static int completionNext(sqlite3_vtab_cursor *cur){
       case COMPLETION_TABLES: {
         if( pCur->pStmt==0 ){
 #if defined(SQLITE_BUILDING_FOR_COMDB2)
-          /* TODO: UNION vtables? */
           sqlite3_prepare_v2(pCur->db, 
                   "SELECT tablename FROM comdb2_tables "
-                  "WHERE tablename NOT LIKE 'sqlite_stat%' ORDER BY 1", 
+                  "WHERE tablename NOT LIKE 'sqlite_stat%' UNION "
+                  "SELECT name FROM comdb2_systables ORDER BY 1",
```
Comment (nirbhayc) [inline sqlite/ext/misc/completion.c]: I gave it a shot but for some reason auto-completion did not return system table names.

File: tools/cdb2sql/cdb2sql.cpp (hunk 2 above)
Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: alpha==@ in my mind. did not want to call it char_at_words[]. I can rename it to char_atglyph_words[] maybe.

File: tools/cdb2sql/cdb2sql.cpp (hunk 3 above)
Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: ok : )

File: sqlite/ext/misc/completion.c (hunk 4 above)
Comment (adizaimi) [reply sqlite/ext/misc/completion.c]: I assume you ran the correct cdb2sql exe. 
It should show after something like 'select * from ^I'  (ie not as the first level token). Ill try to see if it works for me in a fresh checkout/build.

File: tools/cdb2sql/cdb2sql.cpp (hunk 1 above)
Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: sure, but then how do you know to use '@' ?

Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: Should i show '@' as the first word though?

File: sqlite/ext/misc/completion.c (hunk 4 above)
Comment (adizaimi) [reply sqlite/ext/misc/completion.c]: Seems to work for me, try 'select * FROM comdb2_col^I'

Comment (adizaimi) [reply sqlite/ext/misc/completion.c]: Remember, you have to restart the server with the new comdb2 executable to get the autocomplete list from the server itself.

Comment (nirbhayc) [reply sqlite/ext/misc/completion.c]: ok, its working now.

Comment (nirbhayc) [reply sqlite/ext/misc/completion.c]: I think, I used a wrong binary.

File: tools/cdb2sql/cdb2sql.cpp [hunk 5]
```diff
@@ -297,50 +321,72 @@ char *generic_generator(const char *text, int state)
...
 
-    if(endptr == bgn)
-        return rl_completion_matches ((char *) text, &level_one_generator);
+    if (endptr == bgn) { // we were in the middle of the first word
+        if (*bgn == '@')
+            return rl_completion_matches(text, &char_atglyph_generator);
+        else
+            return rl_completion_matches(text, &level_one_generator);
+    }
+    if (bgn < endptr);
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: leftover?

File: tools/cdb2sql/cdb2sql.cpp [hunk 6]
```diff
@@ -175,8 +176,31 @@ const char *words[] = {
...
+  "WITH", NULL,  // must be terminated by NULL
+};
+
+
+const char *char_atglyph_words[] = {
+    "desc", "ls", "send", NULL, // must be terminated by NULL
+};
+
 
+char *char_atglyph_generator (const char *text, int state)
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: static

File: tools/cdb2sql/cdb2sql.cpp [hunk 7]
```diff
@@ -175,8 +176,31 @@ const char *words[] = {
...
+        list_index = 0;
+        len = strlen (text);
+    }
+    while ((name = char_atglyph_words[list_index]) != NULL) {
+        list_index++;
+        if (len == 0 || strncasecmp (name, text, len) == 0) {
+            return strdup (name);
+        }
+    }
+    return ((char *) NULL); // If no names matched, then return NULL.
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: superfluous cast?

File: tools/cdb2sql/cdb2sql.cpp [hunk 8]
```diff
@@ -175,8 +176,31 @@ const char *words[] = {
...
+    static int list_index, len;
+    const char *name;
+    if (!state) { //if state is 0 get the length of text
+        list_index = 0;
+        len = strlen (text);
+    }
+    while ((name = char_atglyph_words[list_index]) != NULL) {
+        list_index++;
+        if (len == 0 || strncasecmp (name, text, len) == 0) {
+            return strdup (name);
```
Comment (nirbhayc) [inline tools/cdb2sql/cdb2sql.cpp]: You don't need strdup() here.

Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: I think i do, the readline api frees the string.

File: tools/cdb2sql/cdb2sql.cpp (hunk 7 above)
Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: This is a separate executable, but ok sure.

File: tools/cdb2sql/cdb2sql.cpp (hunk 6 above)
Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: Separate executable, but why not.

File: tools/cdb2sql/cdb2sql.cpp (hunk 5 above)
Comment (adizaimi) [reply tools/cdb2sql/cdb2sql.cpp]: Thanks, cleaned up.

**Pull request merged**